src/idpi/operators/destagger.py

```diff
@@ -1,39 +1,68 @@
 """algorithm for destaggering a field."""
+# Standard library
+from typing import Literal
+
 # Third-party
+import numpy as np
 import xarray as xr
 
 
-def destagger(field: xr.DataArray, dim: str) -> xr.DataArray:
+def _dstgr(a: np.ndarray):
+    t = a.copy()
+    t[..., 1:] = 0.5 * (a[..., :-1] + a[..., 1:])
+    return t
+
+
+def _dstgr_z(a: np.ndarray):
+    return 0.5 * (a[..., :-1] + a[..., 1:])
+
+
+def destagger(
+    field: xr.DataArray,
+    dim: Literal["x", "y", "generalVertical"],
+) -> xr.DataArray:
     """Destagger a field.
 
-    Args:
-        field (xr.DataArray): field to destagger
-        dim (str): dimension, one of {"x", "y", "generalVertical"}
+    Parameters
+    ----------
+    field: xr.DataArray
+        field to destagger
+    dim: str
+        dimension, one of {"x", "y", "generalVertical"}
 
-    Raises:
-        RuntimeError: todo: describe this
+    Raises
+    ------
+    ValueError:
+        Raises ValueError if dim argument is not one of
+        {"x","y","generalVerticalLayer"}.
 
-    Returns:
-        xr.DataArray: destaggered field with dimensions in
-            {"x","y","generalVerticalLayer"}
+    Returns
+    -------
+    xr.DataArray:
+        destaggered field with dimensions in
+        {"x","y","generalVerticalLayer"}
 
     """
+    dims = list(field.sizes.keys())
     if dim == "x" or dim == "y":
-        field_ = field
-        field_[{dim: slice(1, None)}] = (
-            field[{dim: slice(0, -1)}] + field[{dim: slice(1, None)}]
-        ) * 0.5
-        return field_
+        return xr.apply_ufunc(
+            _dstgr,
+            field.reset_coords(drop=True),
+            input_core_dims=[[dim]],
+            output_core_dims=[[dim]],
+        ).transpose(*dims)
     elif dim == "generalVertical":
-        field_k0_ = field[{dim: slice(0, -1)}].assign_coords(
-            {dim: field[{dim: slice(0, -1)}].generalVertical}
-        )
-
-        field_k1_ = field[{dim: slice(1, None)}].assign_coords(
-            {dim: field[{dim: slice(0, -1)}].generalVertical}
-        )
-        return ((field_k0_ + field_k1_) * 0.5).rename(
-            {"generalVertical": "generalVerticalLayer"}
+        return (
+            xr.apply_ufunc(
```

Comment:
can you remind me why is apply_ufunc needed? According to the doc this seems to be used for applying vectorize ufunc functions that are not xarray aware, like no.<OP>. but this is not the case here. 
If the idea is to bypass the alignment and coordinate operation, simply the removal of the coordinates by itself will already have that effect, isnt? What is the benefit of using apply_ufunct?

Reply:
It's a simple wrapper that allows you to work with pure numpy arrays in the ufunc. It takes care of the data alignment and transposing the array such that the dimension that you want to work along is the last one.

Reply:
Using the apply_ufunc approach here improves the performance by a factor 6 on the test data. That alone is enough reason to use it in my opinion.

Reply:
As discussed, the improvement was 6x in the xy directions (compared to the patched version that does a copy) and more like 10x in the z direction.